### PR TITLE
Add automatic weekly snapshot releases with Github Actions

### PR DESCRIPTION
**Pull request by sam-mccall**
See prototype in the clangd-build/empty repository

https://github.com/clangd-build/empty/releases

https://github.com/clangd-build/empty/actions

### CODE REVIEW

File: .github/actions/pick/dist/index.js [hunk 1]
```diff
@@ -0,0 +1 @@
...n}var r="color: "+this.color;t.splice(1,0,r,"color: inherit");var n=0;var o=0;t[0].replace(/%[a-zA-Z%]/g,function(e){if(e==="%%"){return}n++;if(e==="%c"){o=n}});t.splice(o,0,r)}function log(){var e;return(typeof console==="undefined"?"undefined":_typeof(console))==="object"&&console.log&&(e=console).log.apply(e,arguments)}function save(e){try{if(e){t.storage.setItem("debug",e)}else{t.storage.removeItem("debug")}}catch(e){}}function load(){var e;try{e=t.storage.getItem("debug")}catch(e){}if(!e&&typeof process!=="undefined"&&"env"in process){e=process.env.DEBUG}return e}function localstorage(){try{return localStorage}catch(e){}}e.exports=r(548)(t);var n=e.exports.formatters;n.j=function(e){try{return JSON.stringify(e)}catch(e){return"[UnexpectedJSONParseError]: "+e.message}}},984:function(e,t,r){"use strict";var n=r(228);e.exports=function settle(e,t,r){var o=r.config.validateStatus;if(!r.status||!o||o(r.status)){e(r)}else{t(n("Request failed with status code "+r.status,r.config,null,r.request,r))}}},995:function(e){"use strict";e.exports=function bind(e,t){return function wrap(){var r=new Array(arguments.length);for(var n=0;n<r.length;n++){r[n]=arguments[n]}return e.apply(t,r)}}}});
```
Comment (kadircet) [inline .github/actions/pick/dist/index.js]: does this really need to be checked in ? my js isn't so strong, but it also contains strings like
`/usr/local/google/home/sammccall/src/empty/.github/actions/pick/node_modules/last-successful-gh-commit`

File: .github/workflows/periodic.yaml [hunk 2]
```diff
@@ -0,0 +1,50 @@
+# Workflow to create an auto-buildable release periodically.
+#
+# Releases created with automation credentials don't trigger workflows.
+# Therefore a token `secrets.RELEASE_TOKEN` must exist with public_repo scope.
+name: Periodic release
+on:
+  # Run weekly on sunday at 21:37 UTC (arbitrary)
```
Comment (kadircet) [inline .github/workflows/periodic.yaml]: why not 13:37 :( ?

File: .github/workflows/periodic.yaml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  schedule:
+  - cron: '37 21 * * SUN'
+  # Allow triggering manually:
+  # curl -XPOST -d '{"event_type":"periodic"}' \
+  #  "-HAuthorization: Bearer <token>" \
+  #   https://api.github.com/repos/clangd-build/empty/dispatches
+  repository_dispatch: { types: periodic }
```
Comment (kadircet) [inline .github/workflows/periodic.yaml]: the [documentation](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/events-that-trigger-workflows#example) says: 
`This event does not have additional activity types so it does not support the types keyword.`

are we sure this is legit?

File: .github/workflows/periodic.yaml [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+      env: { GITHUB_TOKEN: "${{ secrets.RELEASE_TOKEN }}" }
+      with:
+        tag_name: snapshot_${{ env.RELEASE_DATE }}
+        release_name: ${{ env.RELEASE_DATE }} @${{ env.RELEASE_COMMIT_SHORT }}
+        body: |
+          Unstable snapshot of clangd on ${{ env.RELEASE_DATE }}.
+
+          Built from llvm/llvm-project@${{ steps.pick.outputs.sha }}.
+        prerelease: true
+        # It would be nice to use draft releases...
```
Comment (kadircet) [inline .github/workflows/periodic.yaml]: what do you mean by `draft releases` exactly?

File: .github/actions/pick/dist/index.js (hunk 1 above)
Comment (sam-mccall) [reply .github/actions/pick/dist/index.js]: Nothing will fetch dependencies for you (here, last-successful-gh-commit as well as actions/core).

So the alternatives are checking in the whole node_modules directory, or bundling it all up into one file as we do here. Using `ncc` is recommended by the actions tutorials, it's a bit sad that it leaks the local paths in.

I added `build.sh` which produces dist/index.js. If you prefer checking in node_modules, I'm also happy to do that.

File: .github/workflows/periodic.yaml (hunk 2 above)
Comment (sam-mccall) [reply .github/workflows/periodic.yaml]: Heh, habit of avoiding popular scheduling times like 00:00 or 13:37 ;-)

File: .github/workflows/periodic.yaml (hunk 3 above)
Comment (sam-mccall) [reply .github/workflows/periodic.yaml]: Docs are out of date, I've tested and it works (iff you send the matching event_type).

Here's the doc for the request: https://developer.github.com/v3/repos/#create-a-repository-dispatch-event
> event_type | string | Required: A custom webhook event name.

File: .github/workflows/periodic.yaml (hunk 4 above)
Comment (sam-mccall) [reply .github/workflows/periodic.yaml]: Releases have a "draft" state, for the following workflow.

1. create a release with draft: true, a description etc
2. upload assets etc
3. publish release (i.e set draft: false)

The release is only visible (to non-owners) when draft is false. (Also the tag is only created when draft is set to false).

However draft releases don't fire any events that the `autobuild.yaml` workflow can trigger on. https://help.github.com/en/actions/automating-your-workflow-with-github-actions/events-that-trigger-workflows#webhook-events:
> Note: The release event is not triggered for draft releases.

So instead, this workflow creates a *non*-draft release, which triggers the `autobuild` workflow, and that workflow immediately marks it as draft again (hiding it), builds/attaches artifacts, and then publishes again.

Added a comment explaining more.

File: .github/workflows/autobuild.yaml [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+  finalize:
+    runs-on: ubuntu-latest
+    needs: build
+    steps:
+    - name: Publish release
+      run: >
+        curl -XPATCH
+        "-HAuthorization: Bearer ${{ secrets.RELEASE_TOKEN }}"
+        "https://api.github.com/repos/${{ github.repository }}/releases/${{ github.event.release.id }}"
+        "-d" '{"draft": false}'
```
Comment (hokein) [inline .github/workflows/autobuild.yaml]: maybe set `prerelease` to `true`?

Comment (sam-mccall) [reply .github/workflows/autobuild.yaml]: By not setting `prerelease`, we preserve the value set when the release was originally created.

**Comment by sam-mccall** [comment]
> Some suggestions:
> 
> * would be nicer to have some documentations (in the PR description?) describing the whole picture of the workflow
I put this in `releases.md`.

> * the release is under pre-release (https://github.com/clangd-build/empty/releases), could we make it `release`?
Only the `periodic` workflow sets the prerelease flag. The automatic snapshot releases are marked as pre-release, which I think is appropriate given that we have no idea about their stability.
We can create manual releases for whichever revision we want (e.g. matching LLVM releases) and make these "full" releases.

> * regarding the asset names, maybe include the release date as well, like `clangd-mac-20200209.zip`
Done.

> * shall we include the SHA-256 Hashes for the release binaries?
I'm not really sure what this would achieve - would we sign them somehow? How would the source of the hashes be more trustworthy than the files themselves?

> * if something is broken during the release process, will we receive an alert notification?
It seems only the "triggering user" gets the notification. That's the user who creates the release, for periodic releases it's the dummy `clangd-build` user I created.

File: .github/actions/pick/dist/index.js (hunk 1 above)
Comment (sam-mccall) [reply .github/actions/pick/dist/index.js]: Thanks for the offline suggestion to just run `npm`, that's much cleaner.

File: .github/workflows/periodic.yaml [hunk 6]
```diff
@@ -0,0 +1,53 @@
+# Workflow to create an auto-buildable release periodically.
+#
+# Releases created with automation credentials don't trigger workflows.
+# Therefore a token `secrets.RELEASE_TOKEN` must exist with public_repo scope.
```
[the remark this answers is not in the document]
Comment (sam-mccall) [reply .github/workflows/periodic.yaml]: Yes, that's right. I don't think it's worth going into that here though, these workflows assume some knowledge of the automation system.

**Pull request merged**